### ISSUE TITLE
Allow users to hide the display of image creation time

### ISSUE THREAD

**Issue by rudolfmayer** [mapillary/mapillary_issues#1390]
#1112 provided showing the to-the-second-exact time of image creation.

While the exact image time might be useful for some purposes, I feel there are serious privacy concerns with displaying such detailed information on the exact timing of routes a certain user takes, which should not be disclosed to the general public without consent.
Especially together with the weaknesses identified in #986 this would allow creating a "nice" profile of users..

I would thus request allowing the user to suppress such information via a configuration setting. At least, allowing to "revert" to the previous state where only the day was displayed, but ideally to an even coarser level - maybe only the calendar week, or the month (as it is e.g. in Google Streetview).

This affects both the webviewer and the API, as well as other means to view the images.

**Comment by richlv** [comment]
but that would also reduce the usefulness of the images - is that the morning or evening picture ? and what about being able to deduce the time by the shadows ?

personally, i'd prefer a warning in the licence agreement etc and keep on displaying that info

**Comment by rudolfmayer** [comment]
- Yes, usefulness of images might be less - BUT:
 - Do we have a convincing story for actually needing the exact time **displayed** to the general puclic? 
For example, automatically filtering the (likely!) "better" images (i.e. the ones not directly facing the sun) from several images taken of the same spot can still be done on the backend, **without** needing to display the time...
 - If "morning or evening" picture is useful information, we could allow the user to disclose approximative timestamps, such as "morning", "noon", "afternoon" - if that really adds value compared to not having the time available...

- Most important: an image of a previously unmapped place that lacks an exact timespamp is worth more than no image of that place....

- Deducing the time by the shadows is (at least with the technology we have at this moment :-) ) a manual effort, thus not that much an issue; also, it is a difference to **estimate** "it was in the late afternoon" vs. knowing the exact time..

- Specifically mentioning this in the license agreement makes sure that Mapillary as a company is "on the safe side". But it will not make users that would want to contribute, but not want to fully disclose their pattern of movement, upload their images...

**Comment by simonmikkelsen** [comment]
A very good reason to display the time to the public is, that then everybody is made aware that the time  is available. You might think that only the date is shown, and is mapping a lot. Then someone uses the API to extract the timestamp which you did not know was possible.

An estimation of time can be quite accurate. If you can trust the schedule of public transport you got it there. In Denmark a lot of companies have a promotional clock that you would photograph. And using shadows you can actually tell the time very accurate, at least down to the quarter of the hour if someone takes the time to make some measures.

Hiding the the time stamp or reducing it more than removing seconds, will be to tell the user that the information is not available. But it is available. Just in the same way as you can look at my mappings and make a good guess about where I live, where my sons kinder garden is and where I work.

To me the usefulness, like the time filter, is a secondary issue. Hiding the time stamp will be telling people something which is not true. The problem does not go away by hiding the time stamp. It just makes it less visible until you are in trouble.

**Comment by rudolfmayer** [comment]
I agree that it is bad to have the time hidden only from the web, and not the API, thus misleading the "common" user; which was apparently exactly how mapillary behaved until the changes of #1112; misleading to users, and a negative surprise when that was changed. That's why I wrote in my request that it applies to **all means to obtain pictures**, not just the web!

In any case, it can't be reasoned that because the timestamp was always available in the API that this is not a privacy concern.

Your point seems to be that the time can be **inferred** (mostly manually, at least ATM, and even if automated, then with not perfect precision, and at the cost of image processing) in **some** (!) images, e.g. as mentioned those that e.g. have the a clock in the image.
This however is not of the same severity as obtaining the timestamp (to the detail of the second, which you get much less frequent in clocks displayed in the public) in a automatically, machine-processable way requiring very little effort. These are two inference attacks that totally differ on the magnitude of threat.
I don't see this therefore as a valid argument to display the exact time....

**Comment by pbb72** [comment]
Could you give an example where seeing the exact timestamp (as opposed to knowing morning/midday/evening/night) would be a danger to the privacy?

Note that you do not need to detect the time in every image to identify it. It's enough to extract it from one image in a whole sequence. And machine detection of clocks can't be very hard. Look at how Mapillary is already detecting license plates and street signs. (Though you could argue that the clocks should also be blurred, just like license plates.)

Reading the time from shadows is more difficult to do for a machine, but my guess is that 1 in every 10 photos shows shadows that are clear enough to manually infer the time to circa 15 minutes precise.

**Comment by rudolfmayer** [comment]
Regarding the privacy concerns
- If you know the exact time stamp you can obviously create much more detailed "profiles" of a user; and I am not sure whether I would want everyone to know the exact time of when I do a mapping trip, really...
- Of course if you spin it further, you can infer a lot of information also on other people, not only the uploader. Of course that requires multiple images to actually contain a person/car/etc, but at least in smaller places that doesn't seem totally improbable..
- Besides profiling, one could for example very easily detect sequences where you were speeding. You have the coordinates of two or more images, the exact time stamps, that's a very easy computation...

These are the concerns I have; maybe I am the only one, but it is still my concern :-)

Regarding the counter arguments that "this is also possible without the timestamp" - this is **really** not the same thing: it is a much **harder** task, and will **never** achieve the same precision! I work long enough in signal processing (albeit much more in other modalities than image) to have a enough of technical background here.
Sure, image detection algorithms are already good, and will undoubtedly improve, and thus one won't be perfectly protected from some inferences to happen, totally agreed to that. But:
- The example that public clocks would be such a tremendous help seems exaggerated; clocks will be captured in a few images only, and can be easily deleted by the user if she wishes so. Also, detecting the time on a clock is surely more complicated than number plates and traffic signs, which are more standardised; sure, it might be possible in the future, but, it can be rather easily removed; even after uploading, if you detected it, you can hide the image.
- Detecting the time from shadows requires additional background information, such as height of the object that casts the shadow (might be detected from the image itself if it is fully visible there, yes), the day, and information on sunrise/sunset on that day in that location; doesn't sound impossible, but again, **much harder** than just reading the metadata from the API :-)
It also won't work for images taken on a cloudy day...
- You won't be able to detect the exact day (if hidden) from just environmental information...


To summarise: I am not suggesting that this should be applied to **all** images; not even that it is the **default** setting; I would just like to have the option to contribute images, but not to reveal more than I want. And as stated in the beginning, images w/o timestamp metadata are for sure still more useful than no image of the place at all...
So, I don't understand the opposition towards this...

A valid counter-argument against my proposal is for sure that mapillary doesn't want to implement this if I am the only one requesting the feature.. Fine, then the alternative for me is to create a script that before uplading will reset the date to midnight on the first day in e.g. the quarter.. Would work for me, but will in my opinion be worse, because it would give false meta-data, rather than no meta-data....

**Comment by de-vries** [comment]
@rudolfmayer I agree with your privacy concerns. 
In the discussion on #1112 there was the suggestion to create the ability to hide certain sequences from your user activity and thus make them a anonymous contribution. For me that would be a sufficient solution to this privacy concern. Do you agree?

OT: maybe this discussion should be on the forum instead of here.

**Comment by jesolem** [comment]
Hey everyone

Just chiming in here to say that we are listening. Step one is to enable "hide my activity" on the profile pages which will hide the activity tab. @knikel is working on profile remake, including this. (This will not hide in API though.) 

Optionally we could:
1) also remove the username under the photos if you set "hide activity".  
or
2) show only day and not hour under the photos if you set "hide activity".

Thoughts about that?

**Comment by shlublu** [comment]
@jesolem Both options sound good to me, they correspond to different concerns. Is that a either/or?

Thanks for having submitted this and explained the rationales behind, @rudolfmayer - I fully support.

**Comment by rudolfmayer** [comment]
I think we should separate the different privacy issues, as they address different concerns.

- One aspect is hiding discoverability of sequences, for that the "hide activity" should be a on/off setting (default off); that is mostly #986

- Another aspect is hide timing information (this issue), and thus this should be a different setting - someone might be ok to have the sequences discovered, but doesn't want to display the exact time. 
Therefore, this could be a off/on setting, or ideally even more detailed, as in 
exact / day / (week) / month
to allow users to display the exact timestamp (default setting), abstract it to day, (maybe also week), or month,

- "Remove username" could then be another setting, independent from the ones above; default to show the username.

In general, I think that the web & API should behave the same; thus, if you set to hide the activity, I think it should also be hidden in the API. Otherwise you create a slightly harder, but still very easy way to again infer the information you want...

**Comment by simonmikkelsen** [comment]
@jesolem To me the big problem is that removing the information some times gives people the false impression, that the information is always hidden. Both time stamp and user identity can under certain circumstances be identified with a certain precision.

If one a both are hidden, I think the settings page should have a big permanent warning, which tells that the information might be pretty safe from the wife. But if you are trying to hide it from e.g. a private detective or a government agency, it may fail.
In countries which are not as stable as most European countries, that may be important knowledge. Of cause, if Sweden tries to invade Denmark again, I guess the setting may not be respected :-)

I agree that there are use cases where it is enough to just hide it. It may be to avoid a speeding ticket or to hide your activities from the wife or neighbours.

**Comment by rudolfmayer** [comment]
@simonmikkelsen I agree, as I also stated, that the behaviour of web and API should be the same; otherwise, you are suggesting to the average user (that only knows the web interface!) that by changing the option in his profile, his timing information will be protected.

Anyhow, do we have any idea on a timeline when we can expect this to be implemented? Otherwise, I will run my now ready script over my 60k images waiting on my HDD, and upload them with a date set to the first of the month and a fixed time...

**Comment by nicolas17** [comment]
Apart from privacy (not wanting your wife to know your past whereabouts), there might also be a safety concern if someone can infer routines and time/place patterns of Mapillary uploaders. People could figure out that not only every week at a certain time I'm in certain area, but also that I'm there walking with an (expensive?) phone in my hand in picture-taking position and potentially paying more attention to the camera alignment than to my surroundings. I'll let you figure out what that implies.

**Comment by pbb72** [comment]
Hiding or modifying the time stamp will only give a false sense of security. The visibility of clocks, well-defined shadows, public transport and many other elements will make it possible to derive the timestamp. Remember that all photos in a sequence were taken within minutes of eachother, so there is a lot of information available.

I don't think we can make Mapillary responsible for preventing detection of time. That responsibility has to be with the users themselves. For example by taking different routes each time (which would be better for Mapillary anyway), and not making pictures in areas that are sensitive to you, or at least not starting and stopping sequences there.

I believe more in a disclaimer that the user, and not Mapillary, is not responsible for their users own privacy.

**Comment by GITNE** [comment]
I am very happy and appreciate the fact that the community, people from different countries and backgrounds take privacy concerns very seriously, while at the same time providing arguments and proposals to work on solutions. This is a great thing to experience in contrast what mass media often spills into people's homes and minds these days. So, thank you all for that. :smile:

I have been thinking about privacy concerns and repercussions of the publication of photos with time stamps with geo location data too, even before I came across this specific GitHub issue. A lot of concerns voiced here are very reasonable. Let me break them down a bit:

> but that would also reduce the usefulness of the images - is that the morning or evening picture?

Yes, or perhaps it would reduce the usefulness of the photos. However, keep in mind that no consumer of Mapillary photos has any right to nor guarantee to the accuracy of data served by Mapillary. If this kind of information and specific guarantees are important to you as a consumer then you should probably look elsewhere or produce this material yourself. You just cannot force or demand from Mapillary users to give you any guarantees. Please keep in mind that Mapillary is a best effort service.

> and what about being able to deduce the time by the shadows?


> Deducing the time by the shadows is (at least with the technology we have at this moment :-) ) a manual effort, thus not that much an issue; also, it is a difference to estimate "it was in the late afternoon" vs. knowing the exact time..

> An estimation of time can be quite accurate. If you can trust the schedule of public transport you got it there. In Denmark a lot of companies have a promotional clock that you would photograph. And using shadows you can actually tell the time very accurate, at least down to the quarter of the hour if someone takes the time to make some measures.

Yes, deducing the time stamp from shadows, just like with any other photo content, is indeed possible, at least to some extent. This is a risk a Mapillary user is willing to take after having more or less carefully assessed before uploading any photo. Every photo on Mapillary bears privacy concerns. But, this does not mean that all risks, concerns, or permutations thereof are acceptable or unacceptable to every user.

However, any argument based on any existing or imaginable technology (including granularity) an adversary may or may not have is inadmissible in this privacy debate. The level of an adversaries sophistication or amount of resources **in** **this** **context** are **always** unknown, especially in the future. So, from a Mapillary user's point of view this will always be a completely unknown variable in his or her risk and repercussion assessment when sharing data with Mapillary (or anybody else). However, a Mapillary user should know or be aware of what data he or she is sharing, and what level of data protection Mapillary can offer.

> personally, i'd prefer a warning in the license agreement etc and keep on displaying that info

It is not that simple. However, I do agree that warnings or notices should be written into Mapillary's terms of service. But, in practice it is not enough to have legal statements on paper. Life is far more complex than that. Especially for non-tech savvy people. Sure, legally it is every user's sole responsibility, regardless of the level of technical understanding, to decide what data he or she is going to upload to Mapillary. And, even though people might have technical understanding and be aware that they are going to upload more or less sensitive data, they may not always have the means to remove this sensitive data beforehand. Bear in mind that for most of the world's population today, the first contact with computers is via their smartphone. Sure, smartphones are computers (with antennas), not much different to PCs and therefor can transform any data, but they follow a genuinely different design to PCs, so does available software, which ultimately results in different effective and practical use cases. So, even geeks with nothing than a smartphone at their hands will have a very hard time to exclude sensitive data from sharing with Mapillary. These people are left with only two options; either upload and risk all possible consequences (good or bad) or do not participate in this great endeavor of mapping the world with others.
Personally, I do not want to exclude anybody in this endeavor nor give them the impression to feel excluded because of privacy concerns. Especially not because some developers might have been too lazy to implement technical solutions which will help protect their human rights. If technical solutions exist (and they do) or are feasible to implement then we should do that and support Mapillary in building those.

> Most important: an image of a previously unmapped place that lacks an exact timespamp is worth more than no image of that place....

Maybe, maybe not. This totally depends on the requirements or expectations of the consumer. In short, this pseudo argument does not apply ever.

> A very good reason to display the time to the public is, that then everybody is made aware that the time is available. You might think that only the date is shown, and is mapping a lot. Then someone uses the API to extract the timestamp which you did not know was possible.

It is true that just hiding data or parts of data may lead to confusion or misconceptions about the true extent of data available. This is why it so important to educate users, not only Mapillary users but all technology users about what any technology can do to them (or others), good or bad. But before this can happen technology builders and providers have to be open and more specific on what they are doing. In this case, Mapillary needs to be more specific on what kind of data Mapillary does collect, beyond photos and the (geek obvious) time stamp. I say "geek obvious" because I am pretty sure that most users do not even know that a time stamp is stored with every photo, and even if they do they are most likely unable (or unlikely) to asses the risks and repercussions of this fact on their privacy. Nothing undermines trust more than unnecessary secrecy of organizations and legal entities. Sometimes, just having forgotten to inform users about the existence of specific data may be perceived by users (or the general public) as *intended* secrecy after they became aware of it. So, I think it is really in Mapillay's best interest to be more open and be more specific on what data users share when uploading. Of course, this notion applies not only to Mapillary but all organizations dealing with (big) data.
The awareness of the extent of available or stored data at Mapillary should not come from the rendering of a time stamp but from publicly available technical documentation and legal documents (like terms of service).

> Optionally we could:
> 1) also remove the username under the photos if you set "hide activity".
> or
> 2) show only day and not hour under the photos if you set "hide activity".

It is not that simple. However, I am convinced that any photos uploaded to Mapillary should be stored with time stamps in full (down to the smallest) granularity (just like all photo pixels are now before blurs are applied). Why? Because it is a protected archive and the entity operating this archive is Mapillary. Trust has to start somewhere. So, the question is "What data should Mapillary redistribute?" rather than how to render (or display) it.

What Mapillary needs to do to protect its user's privacy, is to give them tools to do so, including a guarantee that Mapillary will do so in all products and services (like web viewer or public APIs) they offer (or will offer in the future) from data the users have uploaded. The first step would be to acknowledge that Mapillary can not asses the risks of privacy concerns and repercussions on the privacy of any particular user but they can give users tools to adjust the amount of data they are willing or able to publicly share. With regards to time stamps, this could mean for some users that they are comfortable with sharing time stamps with a granularity down to a second, others perhaps only down to a month. Only users can make this risk assessment. Why? Because we live in different political and socioeconomic regions of the world. What may be an acceptable risk in one country or region may not be in some other, or vice verse. No "one size fits all" approach can exist here. This notion applies to other publicly shared data too.

> To put it in another way. If some one hides the time stamp when they really need it, they also needs to take a lot of precautions when mapping to avoid revealing the time. Some times they will succeed, but most likely they will fail, because it is extremely hard to anonymize any kind of information.

Anonymizing data completely is hard and errors may always slip in. Yet, this is no excuse for not making a best effort on Mapillary's side. Besides, Mapillary does not guarantee nor is required to provide complete and full anonymization of its users. However, Mapillary is required to make a best effort, like making clear statements on data being collected and providing tools to users to protect their privacy.
Otherwise it saying like: "Oh well, we should stop encrypting our online banking sessions because
a) implementation mistakes will eventually reveal the contents of an online banking session,
b) a resourceful adversary will eventually break that particular encryption,
c) someone with quantum computers will break the encryption in just a fraction of a second in the future, or
d) some technologically advanced aliens already have quantum computers (or better) to break any known encryption to man."

To sum it up:
1. Be more specific on what data is collected
2. Educate on risks and repercussions sharing specific data may have on users
3. Give users options and tools to decide what data they want and or are able to share

Please note the order of things. 3 will not work without going through 1 and 2 first. I am aware that this takes additional time, resources, and money but that's the "price" to be payed in a crowd-sourced effort.

**Comment by GITNE** [comment]
I should add that although "no size to fit all" can be found here, Mapillary would still be obliged to find a sane default, which IMHO should result in something what other services do; month and year or full date only.

**Comment by jesolem** [comment]
Hi everyone. Sorry for the incredibly long delay.

I wanted to make an update on this as I recently realized that it has been rather quiet from our end on this topic. I apologize for our lack of involvement. We have recently been in contact with @rudolfmayer  over email and got a better understanding. 

We will for the time being not make any changes to the timestamp and I don't want to make empty promises at this point. The reason is that we have the timestamp in the APIs and we should be transparent and align that with what you see on the site. 

We will however undertake an internal audit of user privacy and will use the feedback above. It is important to us that people feel safe to contribute images to Mapillary. As the current decision is to keep the timestamp and I have no date for the completion of the internal audit I suggest that we close this issue knowing that this information has been moved to an internal issue. We will of course report back here once we have completed our audit. 

Would everyone feel comfortable with that?

**Comment by jesolem** [comment]
I flagged this for forum import and let's continue discussing there. Thank you everyone for sharing your opinions.
https://forum.mapillary.com/
